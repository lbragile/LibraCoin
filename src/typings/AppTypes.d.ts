import { ACTIONS } from "../enums/AppDispatchActions";

export interface ITransaction {
  to: string;
  from: string;
  amount: number | undefined;
  message: string;
  signature: string;
}

export interface IUser {
  publicKey: string;
  balance: number;
}

export interface IBlock {
  index: number;
  prevHash: string;
  currHash: string;
  transactions: ITransaction[];
  timestamp: number;
  merkleRoot: string;
  valid?: boolean;
}
export interface IState {
  verifiedTrans: ITransaction[];
  selectedTrans: ITransaction[];
  users: IUser[];
  chain: IBlock[];
}

export interface IAction {
  type: ACTIONS;
  payload?: unknown;
}
