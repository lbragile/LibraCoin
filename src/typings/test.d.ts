export declare global {
  namespace NodeJS {
    interface Global {
      state: IState;
      dispatch: React.Dispatch<IAction>;
      spki: string;
      spki_hex: string;
      pkcs8: string;
      pkcs8_hex: string;
      privateKey: CryptoKey;
      publicKey: CryptoKey;
      LibraCoinSHA: string;
      LibraCoinEncode: Uint8Array;
      generateKeyMock: jest.Mock;
      exportKeyMock: jest.Mock;
      digestMock: jest.Mock;
      getRandomValuesMock: jest.Mock;
    }
  }
}
