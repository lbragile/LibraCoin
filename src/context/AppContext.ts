import React from "react";
import { IAction, IState } from "../typings/AppTypes";

type TAppContext = { state: IState; dispatch: React.Dispatch<IAction> } | undefined;

export const AppContext = React.createContext<TAppContext>(undefined);
