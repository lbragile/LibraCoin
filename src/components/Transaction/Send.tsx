import React, { useContext } from "react";

import { Form, Button, InputGroup } from "react-bootstrap";
import { AppContext } from "../../context/AppContext";
import { ACTIONS } from "../../enums/AppDispatchActions";
import { IAction, IState } from "../../typings/AppTypes";

export default function Send(): JSX.Element {
  const { state, dispatch } = useContext(AppContext) as { state: IState; dispatch: React.Dispatch<IAction> };

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    dispatch({ type: ACTIONS.SET_SIGNED, payload: { signed: false } });
    dispatch({ type: ACTIONS.SET_VALIDATED, payload: { validated: false } });
    dispatch({ type: ACTIONS.ADD_VERIFIED_TRANS, payload: { trans: state.wallet.details } });
    dispatch({
      type: ACTIONS.SET_DETAILS,
      payload: { details: { from: "", to: "", amount: (0).toFixed(2), message: "", signature: "" } }
    });
  };

  return (
    <Form aria-label="Send Form" noValidate className="col-12 col-lg-5 my-2 my-lg-0 trans-form" onSubmit={handleSubmit}>
      <InputGroup>
        <InputGroup.Prepend>
          <InputGroup.Text>Receiver Public Key</InputGroup.Text>
        </InputGroup.Prepend>
        <Form.Control
          aria-label="Receiver Public Key"
          name="receiver-pk"
          className="text-truncate"
          type="text"
          value={state.wallet.details.to}
          readOnly
        />
      </InputGroup>

      <Form.Text className="text-muted">Make sure this matches the value used when signing</Form.Text>

      <InputGroup className="my-2">
        <Form.Control
          aria-label="Send Amount"
          name="amount"
          type="number"
          value={state.wallet.details.amount}
          readOnly
        />
        <InputGroup.Append>
          <InputGroup.Text className="rounded-right border-left-0">LC</InputGroup.Text>
        </InputGroup.Append>
      </InputGroup>

      <InputGroup className="mb-2">
        <InputGroup.Prepend>
          <InputGroup.Text>Message</InputGroup.Text>
        </InputGroup.Prepend>
        <Form.Control
          aria-label="Send Message"
          name="msg"
          as="textarea"
          value={state.wallet.details.message}
          rows={4}
          placeholder="optional message..."
          readOnly
        />
      </InputGroup>

      <InputGroup>
        <InputGroup.Prepend>
          <InputGroup.Text>Transaction Signature</InputGroup.Text>
        </InputGroup.Prepend>
        <Form.Control
          aria-label="Transaction Signature"
          name="sig"
          type="text"
          className="text-truncate"
          value={state.wallet.details.signature}
          readOnly
        />
      </InputGroup>

      <Form.Text className="text-muted">Receiver uses this along with your public key to verify transaction</Form.Text>

      <Button
        aria-label="Send Button"
        className="mt-2"
        variant={state.wallet.sent ? "success" : "primary"}
        type="submit"
        disabled={!state.wallet.signed}
        block
      >
        <b>{state.wallet.sent ? "Sent" : "Send"}</b>
      </Button>
    </Form>
  );
}
