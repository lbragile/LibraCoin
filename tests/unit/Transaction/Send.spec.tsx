import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import Send from "../../../src/components/Transaction/Send";
import { AppContext } from "../../../src/context/AppContext";
import { ITransaction } from "../../../src/typings/AppTypes";

const { state, dispatch } = global;

const formDetails: ITransaction = {
  to: "",
  from: state.user.publicKey ?? "",
  amount: undefined,
  message: "",
  signature: ""
};

it("renders correctly", () => {
  const { asFragment } = render(
    <AppContext.Provider value={{ state, dispatch }}>
      <Send
        validated={false}
        signed={false}
        handleSubmit={jest.fn()}
        details={formDetails}
        setValidated={jest.fn()}
        setSigned={jest.fn()}
      />
    </AppContext.Provider>
  );

  expect(screen.getByRole("form", { name: /Send Form/i })).toHaveFormValues({
    "receiver-pk": "",
    amount: null,
    msg: "",
    sig: ""
  });

  const receiverPublicKey = screen.getByRole("textbox", { name: /Receiver Public Key/i });
  expect(receiverPublicKey).toHaveAttribute("readOnly");
  expect(receiverPublicKey).not.toBeRequired();

  expect(screen.getByText(/Make sure this matches/i)).toBeInTheDocument();

  const sendAmount = screen.getByRole("spinbutton", { name: /Send Amount/i });
  expect(sendAmount).toBeDisabled();
  expect(sendAmount).not.toHaveAttribute("placeholder");
  expect(sendAmount).not.toBeRequired();

  const sendMessage = screen.getByRole("textbox", { name: /Send Message/i });
  expect(sendMessage).toHaveAttribute("readOnly");
  expect(sendMessage).toHaveAttribute("placeholder", "optional message...");
  expect(sendMessage).not.toBeRequired();

  const signature = screen.getByRole("textbox", { name: /Transaction Signature/i });
  expect(signature).toHaveAttribute("readOnly");
  expect(signature).not.toBeRequired();

  expect(screen.getByText(/Receiver uses this along/i)).toBeInTheDocument();

  expect(asFragment()).toMatchSnapshot();
});

describe("send button state", () => {
  it("is enabled when signed", () => {
    render(
      <AppContext.Provider value={{ state, dispatch }}>
        <Send
          validated={false}
          signed={true}
          handleSubmit={jest.fn()}
          details={formDetails}
          setValidated={jest.fn()}
          setSigned={jest.fn()}
        />
      </AppContext.Provider>
    );

    expect(screen.getByRole("button")).toBeEnabled();
    expect(screen.getByRole("button")).toHaveTextContent("Send");
  });

  it("is disabled when not signed", () => {
    render(
      <AppContext.Provider value={{ state, dispatch }}>
        <Send
          validated={false}
          signed={false}
          handleSubmit={jest.fn()}
          details={formDetails}
          setValidated={jest.fn()}
          setSigned={jest.fn()}
        />
      </AppContext.Provider>
    );

    expect(screen.getByRole("button")).toBeDisabled();
    expect(screen.getByRole("button")).toHaveTextContent("Send");
  });
});

it("submits form when send button is pressed", () => {
  const handleSubmit = jest.fn().mockImplementation((e) => e.preventDefault());

  render(
    <AppContext.Provider value={{ state, dispatch }}>
      <Send
        validated={false}
        signed={true}
        handleSubmit={handleSubmit}
        details={formDetails}
        setValidated={jest.fn()}
        setSigned={jest.fn()}
      />
    </AppContext.Provider>
  );

  fireEvent.click(screen.getByRole("button"));

  expect(handleSubmit).toHaveBeenCalledTimes(1);
});
