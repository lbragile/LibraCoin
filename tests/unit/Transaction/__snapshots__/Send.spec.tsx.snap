// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <form
    aria-label="Send Form"
    class="col-12 col-lg-5 my-2 my-lg-0 trans-form"
    novalidate=""
  >
    <div
      class="input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Receiver Public Key
        </span>
      </div>
      <input
        aria-label="Receiver Public Key"
        class="text-truncate form-control"
        name="receiver-pk"
        readonly=""
        type="text"
        value=""
      />
    </div>
    <small
      class="text-muted form-text"
    >
      Make sure this matches the value used when signing
    </small>
    <div
      class="my-2 input-group"
    >
      <input
        aria-label="Send Amount"
        class="form-control"
        disabled=""
        name="amount"
        type="number"
        value=""
      />
      <div
        class="input-group-append"
      >
        <span
          class="rounded-right border-left-0 input-group-text"
        >
          LC
        </span>
      </div>
    </div>
    <div
      class="mb-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Message
        </span>
      </div>
      <textarea
        aria-label="Send Message"
        class="form-control"
        name="msg"
        placeholder="optional message..."
        readonly=""
        rows="4"
      />
    </div>
    <div
      class="input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Transaction Signature
        </span>
      </div>
      <input
        aria-label="Transaction Signature"
        class="form-control"
        name="sig"
        readonly=""
        type="text"
        value=""
      />
    </div>
    <small
      class="text-muted form-text"
    >
      Receiver uses this along with your public key to verify transaction
    </small>
    <button
      aria-label="Send Button"
      class="mt-2 btn btn-primary btn-block"
      disabled=""
      type="submit"
    >
      <b>
        Send
      </b>
    </button>
  </form>
</DocumentFragment>
`;
